Allow proxy to work on both ruby 1.9 and 1.8

GregSutcliffe opened the pull request:
Fixes Bug #1680

I;d like some more thorough testing on Ruby 1.8 and 1.9 - it seemed to work for my two proxies in those environments. I tested creating a host to ensure the discover-ip-by-ping stuff still works, and I imported a subnet to ensure the changes there work in both rubies. Still, more formal tests would be good ;)

Commit:
Changes for Ruby 1.8/9 compatibilty, and Debian packaging guidelines

* Allow proxy to work on both ruby 1.9 and 1.8
* Require ping gem on ruby1.8
* Handle public folder changes in Sinatra
* Don't load rubygems if USE_GEMS=false specified, to comply with Debian guidelines
* Add USE_GEMS to the initscript

ohadlevy commented on bin/smart-proxy:
```diff
@@ -18,7 +24,7 @@ class SmartProxy < Sinatra::Base
   set :root, APP_ROOT
   set :views, APP_ROOT + '/views'
-  set :public, APP_ROOT + '/public'
+  set :public_folder, APP_ROOT + '/public'
```
do you know which version of sinatra changes this option? I'm concern it might break older package based deployments

GregSutcliffe replied:
Not sure. It worked on my machine, but thats quite recent. I suppose we could leave this out for now - it only generates a warning on 1.9

ohadlevy commented on lib/proxy/dhcp/server/isc.rb:
```diff
@@ -162,6 +162,7 @@ def omcmd cmd, msg=nil
     end
 
     def report msg, response=""
+      response = [*response].compact
```
hmm.. whats the meaning of this one? it would convert it to an array (which was not there before)

GregSutcliffe replied:
Something about how we use the response object has changed in Ruby1.9 - if I leave this out I get an error:

E, [2012-06-13T13:05:26.511526 #21978] ERROR -- : undefined method `grep' for "":String

I actually wrote this line a while back, so details are hazy, but I _think_ I tried catching an empty string (something like "if response.nil? or response == "" or !response.grep(..." but it didn't seem to work. If you can get around it that way, go for it :)

GregSutcliffe replied:
Ah I recall. If you test for response.empty? then the if test (165) passes, but the logger line (166) fails because it tries to perform "".join. Rather than change a whole ton of code, the above fix seemed quicker.

GregSutcliffe replied:
I also tried defining the method as "response=nil" but that also doesn't seem to behave correctly

GregSutcliffe replied:
Ok, my leases file was the cause, not the code. new patch incoming.

ohadlevy commented on lib/proxy/dhcp/subnet.rb:
```diff
@@ -1,6 +1,5 @@
 require 'ipaddr'
 require 'proxy/dhcp/monkey_patches' unless IPAddr.new.respond_to?('to_range')
-require 'ping'
```
easy fix ;) if you do that, then you also need to remove that relevant code.

note that it was a needed for windows, so if we remove that bit, we need an alternative way for the proxy to ping on windows (i'm guessing executing ping would work, just need to figure out the arg list).

GregSutcliffe replied:
I removed it because it doesn't exist in ruby1.9 at all - not even as a gem. You can find some third-party ports, but that all feels a bit hacky. Since it didn't seem to affect 1.8 either, I didn't worry. I have no windows machines to test on, though, so I'm unsure how to fix the problem.

We are still including 'net/ping' - is there overlap?

ohadlevy replied:
good question, I'm not sure.

but I dont mind removing this gem, as we dont run as root by default, so reverting to ping command should be fine.

we were thinking of doing more stuff (e.g. using arp, tcp ping etc) so maybe we should consider using nmap or something instead

GregSutcliffe replied:
Sounds good, but outside the scope of this pull request :)

Are you happy to merge as is, or shall I add some kind of "require 'ping' if RUBY_VERSION < 1.9" ?

GregSutcliffe commented on lib/proxy/dhcp/server/isc.rb:
```diff
@@ -162,7 +162,7 @@ def omcmd cmd, msg=nil
 
     def report msg, response=""
-      if response.nil? or !response.grep(/can't|no more|not connected|Syntax error/).empty?
+      if response.nil? or (!response.empty? and !response.grep(/can't|no more|not connected|Syntax error/).empty?)
```
Does that look better @ohadlevy ?

ohadlevy replied:
I'm not sure, but does empty? works on strings too? (I know it works in rails, but i think thats rails specific)

GregSutcliffe replied:
yep:

[gsutcliffe@osiris modules]$ irb
irb(main):001:0> str=""
=> ""
irb(main):002:0> str.empty?
=> true

ohadlevy commented:
@GregSutcliffe can you rebase, so it would easier to understand whats pending and whats not?
thanks

GregSutcliffe commented:
@ohadlevy rebased. Outstanding issues seem to be the "require ping" and the "public" vs "public_folder". Happy to wrap both of those in a version test if you want.

ohadlevy commented:
@GregSutcliffe did you rebase and pushed it? the public_folder is a sinatra version thing, so that should probably depend on that, the require ping makes sense, thanks!

GregSutcliffe commented:
I've rebased, added the changes, and added the necessary rubygems patches for Debian compatibility. Be good if we could merge this and then tag as foreman-proxy_1.0rc1 :)

I've also tested on 1.9 and 1.8, and it seems to work as intended

ohadlevy commented on bin/smart-proxy:
```diff
@@ -3,7 +3,18 @@
+end
+
+# Don't load rubygems in production, it's for developers
+PROD_MODE = [ "true", "t", true ].include? ENV['PRODUCTION_MODE'] ? true : false
```
I'm not sure I get this, what if sinatra is distributed as a gem?

GregSutcliffe replied:
The idea is that of least-surprise - if you just run "smart-proxy" it requires rubygems. Debian deosn;t like that though, so now you can run "PRODUCTION_MODE=true smart-proxy" and it won't require rubygems. The init script uses this for Debian compliance.

I guess normally you'd have "DEVELOPER_MODE" but I didn't want to break anything people we're already doing.

ohadlevy commented:
tests are failing..

<pre>

/usr/bin/ruby -I"lib:lib:test" -I"/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib" "/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/rake_test_loader.rb" "test/puppetca_test.rb" "test/subnet_test.rb" "test/server_test.rb" "test/proxy_test.rb" "test/puppet_environment_test.rb" "test/tftp_test.rb" "test/puppet_class_test.rb" "test/util_test.rb" "test/record_test.rb" 
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/ext/module.rb:36:in `const_missing': uninitialized constant Proxy::USE_GEMS (NameError)
    from ./lib/proxy.rb:8
    from ./test/test_helper.rb:4:in `require'
    from ./test/test_helper.rb:4
    from /home/olevy/git/proxy/test/puppetca_test.rb:1:in `require'
    from /home/olevy/git/proxy/test/puppetca_test.rb:1
    from /usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/rake_test_loader.rb:15:in `require'
    from /usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/rake_test_loader.rb:15
    from /usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/rake_test_loader.rb:4:in `select'
    from /usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/rake_test_loader.rb:4
rake aborted!
Command failed with status (1): [/usr/bin/ruby -I"lib:lib:test" -I"/usr/lib...]

Tasks: TOP => default => test
(See full trace by running task with --trace)
olevy@super:~/git/proxy(develop)$ rake --trace
WARNING: 'require 'rake/rdoctask'' is deprecated.  Please use 'require 'rdoc/task' (in RDoc 2.4.2+)' instead.
    at /usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/rdoctask.rb
rake/gempackagetask is deprecated.  Use rubygems/package_task instead
** Invoke default (first_time)
** Invoke test (first_time)
** Execute test
/usr/bin/ruby -I"lib:lib:test" -I"/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib" "/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/rake_test_loader.rb" "test/puppetca_test.rb" "test/subnet_test.rb" "test/server_test.rb" "test/proxy_test.rb" "test/puppet_environment_test.rb" "test/tftp_test.rb" "test/puppet_class_test.rb" "test/util_test.rb" "test/record_test.rb" 
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/ext/module.rb:36:in `const_missing': uninitialized constant Proxy::USE_GEMS (NameError)
    from ./lib/proxy.rb:8
    from ./test/test_helper.rb:4:in `require'
    from ./test/test_helper.rb:4
    from /home/olevy/git/proxy/test/puppetca_test.rb:1:in `require'
    from /home/olevy/git/proxy/test/puppetca_test.rb:1
    from /usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/rake_test_loader.rb:15:in `require'
    from /usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/rake_test_loader.rb:15
    from /usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/rake_test_loader.rb:4:in `select'
    from /usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/rake_test_loader.rb:4
rake aborted!
Command failed with status (1): [/usr/bin/ruby -I"lib:lib:test" -I"/usr/lib...]
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/file_utils.rb:53:in `create_shell_runner'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/file_utils.rb:45:in `call'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/file_utils.rb:45:in `sh'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/file_utils_ext.rb:39:in `sh'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/file_utils.rb:82:in `ruby'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/file_utils_ext.rb:39:in `ruby'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/testtask.rb:99:in `define'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/file_utils_ext.rb:60:in `verbose'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/testtask.rb:98:in `define'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:205:in `call'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:205:in `execute'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:200:in `each'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:200:in `execute'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:158:in `invoke_with_call_chain'
/usr/lib/ruby/1.8/monitor.rb:242:in `synchronize'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:151:in `invoke_with_call_chain'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:176:in `invoke_prerequisites'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:174:in `each'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:174:in `invoke_prerequisites'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:157:in `invoke_with_call_chain'
/usr/lib/ruby/1.8/monitor.rb:242:in `synchronize'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:151:in `invoke_with_call_chain'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/task.rb:144:in `invoke'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/application.rb:116:in `invoke_task'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/application.rb:94:in `top_level'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/application.rb:94:in `each'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/application.rb:94:in `top_level'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/application.rb:133:in `standard_exception_handling'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/application.rb:88:in `top_level'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/application.rb:66:in `run'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/application.rb:133:in `standard_exception_handling'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/lib/rake/application.rb:63:in `run'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.2.2/bin/rake:33
/home/olevy/.gem/bin/rake:19:in `load'
/home/olevy/.gem/bin/rake:19
Tasks: TOP => default => test
</pre>

GregSutcliffe commented:
Ok, I'd forgotten that the libraries can be directly required by the tests. I've moved the ruby environment checks to it's own file, and required that file wherever we're testing those variables. Tests seem ok now.

ohadlevy commented:
merged
